src/java.base/share/classes/java/util/regex/Pattern.java

```diff
@@ -1095,6 +1096,9 @@ public static Pattern compile(String regex) {
      * Compiles the given regular expression into a pattern with the given
      * flags.
      *
+     * <p>Setting {@link #CANON_EQ} among the flags may impose a moderate risk
```

Comment:
This may be a candidate for `@apiNote`, same thing for the note added to CANON_EQ.

Reply:
The choice of a `<p>` paragraph rather than `@apiNote` is for consistency with similar commentary paragraphs in the specs of `CASE_INSENSITIVE`, `UNICODE_CASE`, and `UNICODE_CHARACTER_CLASS`.

I have no problems in using `@apiNote` instead, but then it would be better to apply the same for the other mentioned flags as well.

Reply:
Okay, I see your point and to use apiNote consistently would require "converting" some of the existing text to apiNote too.

I'm still mulling over Pattern.compile throwing OOME. An implNote is probably the right category for this, in which case it can start with "In the the JDK Reference Implementation ...".  I assume the static Pattern.matches needs same, and also Pattern.matcher for the lazy compilation case.

Reply:
There's no hard limit for the number of combining marks in the Unicode specification, even though in practice it never reaches the implementation limit. A high number of combining marks is thus more akin to a a resource exhaustion condition than to anything else, IMO.

Even today, compilation of a pattern risks throwing an OOME anyway when trying to generate the permutations. Pre-emptively throwing an OOME just anticipates and extrapolates this behavior beyond the `int` limit of array lengths.

Alternatively, compilation (greedy or lazy) could throw `PatternSyntaxException`, although there is not really something wrong with syntax.

I'll add `@implNote` to the other methods you mention.

Reply:
The CSR will be updated once this PR stabilizes.

Reply:
@AlanBateman

I had previously discussed with @rgiulietti whether OOME or PatternSyntaxException was more appropriate. The issue is that you might try to compile a pattern that contains a character with N combining diacritics, and it might work fine. But if you change that character to have N+1 combining diacritics, it might throw OOME. There's no syntax difference, but rather the issue is hitting an internal implementation limit.

There's a bit of a precedent for throwing OOME in such cases. Various places in the library try to grow arrays. If the required array size is greater than MAX_VALUE, the library pre-emptively throws OOME without even trying to allocate the array.